spec/support/mock_intake.rb

```diff
@@ -187,3 +187,116 @@ def print_received
 
   config.include MockIntake::WaitFor, :mock_intake
 end
+
+class EventCollector
+  class << self
+    def method_missing(name, *args, &block)
+      instance.send(name, *args, &block)
+    end
+
+    def instance
+      @instance ||= new
+    end
+  end
+
+  attr_reader(
+      :errors,
+      :metadatas,
+      :metricsets,
+      :requests,
+      :spans,
+      :transactions
+  )
+
+  def initialize
+    clear!
+  end
+
+  def catalog(json)
+    case json.keys.first
+    when 'transaction' then transactions << json.values.first
+    when 'span' then spans << json.values.first
+    when 'error' then errors << json.values.first
+    when 'metricset' then metricsets << json.values.first
+    when 'metadata' then metadatas << json.values.first
+    end
+  end
+
+  def clear!
+    @requests = []
+
+    @errors = []
+    @metadatas = []
+    @metricsets = []
+    @spans = []
+    @transactions = []
+  end
+
+  def transaction_metrics
+    metricsets.select { |set| set['transaction'] && !set['span'] }
+  end
+
+  def span_metrics
+    metricsets.select { |set| set['transaction'] && set['span'] }
+  end
+
+  def wait_for(timeout: 5, **expected)
+
+    if expected.empty? && !block_given?
+      raise ArgumentError, 'Either args or block required'
+    end
+
+    Timeout.timeout(timeout) do
+      loop do
+        sleep 0.01
+
+        missing = expected.reduce(0) do |total, (kind, count)|
+          total + (count - send(kind).length)
+        end
+
+        next if missing > 0
+
+        unless missing == 0
+          puts format(
+                   'Expected %s. Got %s',
+                   expected,
+                   missing
+               )
+          print_received
+        end
+
+        if block_given?
+          next unless yield(self)
+        end
+
+        break true
+      end
+    end
+  rescue Timeout::Error
+    puts format('Died waiting for %s', block_given? ? 'block' : expected)
+    puts '--- Received: ---'
+    print_received
+    raise
+  end
+
+  private
+
+  def print_received
+    pp(
+        transactions: transactions.map { |o| o['name'] },
+        spans: spans.map { |o| o['name'] },
+        errors: errors.map { |o| o['culprit'] },
+        metricsets: metricsets,
+        metadatas: metadatas.count
+    )
+  end
+end
+
+shared_context 'event_collector' do
+  before do
+    EventCollector.clear!
+  end
+  after do
+    EventCollector.clear!
+  end
```

Comment:
Is this to be _absolutely_ sure it's empty? 😅

Reply:
🧹